backends-clickhouse/src/main/scala/org/apache/spark/sql/delta/catalog/ClickHouseTableV2.scala

```diff
@@ -0,0 +1,279 @@
+  def createV1Relation(
+      partitionFilters: Seq[Expression] = Nil,
+      snapshotToUseOpt: Option[Snapshot] = None,
+      isTimeTravelQuery: Boolean = false,
+      cdcOptions: CaseInsensitiveStringMap = CaseInsensitiveStringMap.empty): BaseRelation = {
+    val snapshotToUse = snapshotToUseOpt.getOrElse(DeltaLogAdapter.snapshot(deltaLog))
+    if (snapshotToUse.version < 0) {
+      // A negative version here means the dataPath is an empty directory. Read query should error
+      // out in this case.
+      throw DeltaErrors.pathNotExistsException(deltaLog.dataPath.toString)
+    }
+    val fileIndex =
+      new TahoeLogFileIndex(spark, deltaLog, deltaLog.dataPath, snapshotToUse, partitionFilters)
+    val fileFormat: DeltaMergeTreeFileFormat = getFileFormat(getMetadata)
+    new HadoopFsRelation(
+      fileIndex,
+      partitionSchema =
+        DeltaColumnMapping.dropColumnMappingMetadata(snapshotToUse.metadata.partitionSchema),
+      // We pass all table columns as `dataSchema` so that Spark will preserve the partition column
+      // locations. Otherwise, for any partition columns not in `dataSchema`, Spark would just
+      // append them to the end of `dataSchema`
+      dataSchema = DeltaColumnMapping.dropColumnMappingMetadata(
+        ColumnWithDefaultExprUtils.removeDefaultExpressions(
+          SchemaUtils.dropNullTypeColumns(snapshotToUse.metadata.schema))),
+      bucketSpec = bucketOption,
+      fileFormat,
+      // `metadata.format.options` is not set today. Even if we support it in future, we shouldn't
+      // store any file system options since they may contain credentials. Hence, it will never
+      // conflict with `DeltaLog.options`.
+      snapshotToUse.metadata.format.options ++ options
+    )(
+      spark
+    ) with InsertableRelation {
+      def insert(data: DataFrame, overwrite: Boolean): Unit = {
+        throw new UnsupportedOperationException()
+//        val mode = if (overwrite) SaveMode.Overwrite else SaveMode.Append
+        // Insert MergeTree data through DataSource V1
+      }
+    }
+  }
+
+  def getFileFormat(meta: Metadata): DeltaMergeTreeFileFormat = {
+    val fileFormat = new DeltaMergeTreeFileFormat(
+      meta,
+      dataBaseName,
+      tableName,
+      Seq.empty[Attribute],
+      orderByKeyOption,
+      primaryKeyOption,
+      clickhouseTableConfigs,
+      partitionColumns)
+    fileFormat
+  }
+}
+
+object ClickHouseTableV2 extends Logging {
+  val deltaLog2Table = mutable.HashMap[DeltaLog, ClickHouseTableV2]()
+
+  def partsPartitions(
+      deltaLog: DeltaLog,
+      relation: HadoopFsRelation,
+      selectedPartitions: Array[PartitionDirectory],
+      output: Seq[Attribute],
+      bucketedScan: Boolean,
+      optionalBucketSet: Option[BitSet],
+      optionalNumCoalescedBuckets: Option[Int],
+      disableBucketedScan: Boolean): Seq[InputPartition] = {
+    val tableV2 = ClickHouseTableV2.deltaLog2Table(deltaLog)
```

Comment:
it seems it can not use this way to get the `ClickHouseTableV2`, because if there is no writing data operation in this spark session, right?

Reply:
fixed